File: frontend/demo/component/inputfields/input-field-helper.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/horizontal-layout';
import '@vaadin/icon';
import '@vaadin/icons';
import '@vaadin/password-field';
import '@vaadin/text-field';
import type { PasswordFieldValueChangedEvent } from '@vaadin/password-field';
import { applyTheme } from 'Frontend/generated/theme';

type PasswordStrength = 'moderate' | 'strong' | 'weak';

const StrengthColor: Record<PasswordStrength, string> = {
  weak: 'var(--lumo-error-color)',
  moderate: '#e7c200',
  strong: 'var(--lumo-success-color)',
};

@customElement('input-field-helper')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private strengthText: PasswordStrength = 'weak';

  @state()
  private strengthColor = StrengthColor.weak;

  protected override render() {
    return html`
      <vaadin-horizontal-layout theme="spacing">
        <!-- tag::snippet[] -->
        <vaadin-text-field
          label="Phone number"
          helper-text="Include country and area prefixes"
        ></vaadin-text-field>

        <vaadin-password-field
          label="Password"
          @value-changed="${this.onPasswordChanged}"
          reveal-button-hidden
        >
          <vaadin-icon
            icon="vaadin:check"
            slot="suffix"
            style="color:${StrengthColor.strong}"
            ?hidden="${this.strengthText !== 'strong'}"
          ></vaadin-icon>

          <div slot="helper">
            Password strength:
            <span style="color:${this.strengthColor}">${this.strengthText}</span>
          </div>
        </vaadin-password-field>
        <!-- end::snippet[] -->
      </vaadin-horizontal-layout>
    `;
  }

  private onPasswordChanged(event: PasswordFieldValueChangedEvent) {
    let strength: PasswordStrength = 'weak';
    const { value } = event.detail;
    if (value) {
      if (value.length > 9) {
        strength = 'strong';
      } else if (value.length > 5) {
        strength = 'moderate';
      }
    }
    this.strengthText = strength;
    this.strengthColor = StrengthColor[strength];
  }
}

File: src/main/java/com/vaadin/demo/component/inputfields/InputFieldHelper.java
package com.vaadin.demo.component.inputfields;

import com.vaadin.flow.component.Text;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.textfield.PasswordField;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.value.ValueChangeMode;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("input-field-helper")
public class InputFieldHelper extends HorizontalLayout {

  private Icon checkIcon;
  private Span passwordStrengthText;

  public InputFieldHelper() {
    // tag::snippet[]
    TextField textField = new TextField();
    textField.setLabel("Phone number");
    textField.setHelperText("Include country and area prefixes");

    PasswordField passwordField = new PasswordField();
    passwordField.setLabel("Password");
    passwordField.setRevealButtonVisible(false);

    checkIcon = VaadinIcon.CHECK.create();
    checkIcon.setVisible(false);
    passwordField.setSuffixComponent(checkIcon);

    Div passwordStrength = new Div();
    passwordStrengthText = new Span();
    passwordStrength.add(new Text("Password strength: "), passwordStrengthText);
    passwordField.setHelperComponent(passwordStrength);

    add(textField);
    add(passwordField);
    // end::snippet[]

    passwordField.setValueChangeMode(ValueChangeMode.EAGER);
    passwordField.addValueChangeListener(e -> {
      String password = e.getValue();
      updateHelper(password);
    });

    updateHelper("");
  }

  private void updateHelper(String password) {
    if (password.length() > 9) {
      passwordStrengthText.setText("strong");
      passwordStrengthText.getStyle().set("color", "var(--lumo-success-color)");
      checkIcon.setVisible(true);
    } else if (password.length() > 5) {
      passwordStrengthText.setText("moderate");
      passwordStrengthText.getStyle().set("color", "#e7c200");
      checkIcon.setVisible(false);
    } else {
      passwordStrengthText.setText("weak");
      passwordStrengthText.getStyle().set("color", "var(--lumo-error-color)");
      checkIcon.setVisible(false);
    }
  }

  public static class Exporter extends DemoExporter<InputFieldHelper> { // hidden-source-line
  } // hidden-source-line
}
